### PR TITLE
[Layout] Move [ASLayoutSpec children] from std::map to NSMutableArray

### PR DESCRIPTION
**Pull request by maicki**
Currently we have a pretty bad overhead accessing the children property in `ASLayoutSpec` as it always converts the `std::map` to an `NSArray` if the children getter is called. Let's get rid of that

### TODO:
- [x] Introduce `ASNullLayoutSpec` that basically act as a `Null Object`
- [x] Only expose `setChild:atIndex:` to subclasses of `ASLayoutSpec` for internal use.
- [x] Add fast enumeration to `ASLayoutable` to quickly iterate through `children`
- [x] Create singleton for `ASNullLayoutSpec`

### CODE REVIEW

**Comment by Adlai-Holler** [comment]
LGTM

**Comment by maicki** [comment]
@appleguy @levi Could I please get another review over it?

File: AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  ASDisplayNodeAssert(self.isMutable, @"Cannot set properties when layout spec is not mutable");
+  
+  id<ASLayoutable> layoutable = child ? [self layoutableToAddFromLayoutable:child] : [[ASNullLayoutSpec alloc] init];
+  
+  if (child) {
+    if (_childrenArray.count < index) {
+      // Fill up the array with null objects until index to fill gaps
+      NSInteger i = _childrenArray.count;
+      while (i < index) {
+        _childrenArray[i] = [[ASNullLayoutSpec alloc] init];
```
Comment (appleguy) [inline AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm]: Let's make this a singleton, yay we have our own kCFNull!

File: AsyncDisplayKit/Layout/ASLayoutSpec.h [hunk 2]
```diff
@@ -81,12 +60,9 @@ NS_ASSUME_NONNULL_BEGIN
...
 
-/**
- * Returns the child added to this layout spec using the given index.
- *
- * @param index An identifier associated withe the child.
- */
-- (nullable id<ASLayoutable>)childForIndex:(NSUInteger)index;
+@end
+
+@interface ASNullLayoutSpec : ASLayoutSpec
```
Comment (appleguy) [inline AsyncDisplayKit/Layout/ASLayoutSpec.h]: If we use a singleton, is it possible to eliminate this?  Maybe we could set the singleton to mutable = NO?  It's OK to keep this if having a singleton of a separate class type is valuable (in which case you can add a + method here).

Does it need to be in the public header / what's the main use case for developers?

File: AsyncDisplayKitTests/ASDisplayNodeTests.m [hunk 3]
```diff
@@ -2048,4 +2048,24 @@ - (void)testThatOnDidLoadWorks
   XCTAssertEqualObjects(calls, expected);
 }
 
+- (void)testSubnodesFastEnumeration
```
Comment (appleguy) [inline AsyncDisplayKitTests/ASDisplayNodeTests.m]: Yay, thanks for adding a test!

**Comment by appleguy** [comment]
@maicki you're cleared to merge this.  I wasn't able to do a line-for-line review of the logic, but I trust your consistently high attention to detail :).  After your final changes, let's make sure to run this through the Pinterest testing gauntlet too.  Since we're quite confident in it and it passes unit tests, you could land to master and follow up with Hannah to get testing on it in the following day or two.

**Comment by maicki** [comment]
@appleguy @levi Ready for further review.

File: AsyncDisplayKit/Layout/ASLayoutSpec.h (hunk 2 above)
Comment (maicki) [inline AsyncDisplayKit/Layout/ASLayoutSpec.h]: Let the singleton still be separate class but moved it out of a public header.

File: AsyncDisplayKit/Layout/ASLayoutSpec.mm [hunk 4]
```diff
@@ -162,67 +114,63 @@ - (void)setParent:(id<ASLayoutable>)parent
   }
 }
 
+
+#pragma mark - Child
+
 - (void)setChild:(id<ASLayoutable>)child
 {
-  ASDisplayNodeAssert(self.isMutable, @"Cannot set properties when layout spec is not mutable");
+  ASDisplayNodeAssert(self.isMutable, @"Cannot set properties when layout spec is not mutable");;
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/Layout/ASLayoutSpec.mm]: Remove extra semicolon

File: AsyncDisplayKit/Layout/ASLayoutSpec.mm [hunk 5]
```diff
@@ -162,67 +114,63 @@ - (void)setParent:(id<ASLayoutable>)parent
...
 {
-  ASDisplayNodeAssert(self.isMutable, @"Cannot set properties when layout spec is not mutable");
-  if (child) {
-    id<ASLayoutable> finalLayoutable = [self layoutableToAddFromLayoutable:child];
-    _children[index] = finalLayoutable;
-  } else {
-    _children.erase(index);
+  ASDisplayNodeAssert(_childrenArray.count < 2, @"This layout spec does not support more than one child. Use the setChildren: or the setChild:AtIndex: API");
+  
+  if (_childrenArray.count) {
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/Layout/ASLayoutSpec.mm]: Use `return _childrenArray.firstObject`

File: AsyncDisplayKit/Layout/ASLayoutSpec.mm [hunk 6]
```diff
@@ -162,67 +114,63 @@ - (void)setParent:(id<ASLayoutable>)parent
...
-- (void)setChild:(id<ASLayoutable>)child forIndex:(NSUInteger)index
+- (id<ASLayoutable>)child
 {
-  ASDisplayNodeAssert(self.isMutable, @"Cannot set properties when layout spec is not mutable");
-  if (child) {
-    id<ASLayoutable> finalLayoutable = [self layoutableToAddFromLayoutable:child];
-    _children[index] = finalLayoutable;
-  } else {
-    _children.erase(index);
+  ASDisplayNodeAssert(_childrenArray.count < 2, @"This layout spec does not support more than one child. Use the setChildren: or the setChild:AtIndex: API");
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/Layout/ASLayoutSpec.mm]: Should we remove this `-child` API completely? Obviously it's not critical but it seems silly to make this big distinction between elements that support 1 children and elements that support many, unless I'm missing something.

File: AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.h [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+ * property that behind the scenes is calling setChild:forIndex:.
+ */
+- (void)setChild:(id<ASLayoutable>)child forIndex:(NSUInteger)index;
+
+/**
+ * Returns the child added to this layout spec using the given index.
+ *
+ * @param index An identifier associated with the the child.
+ */
+- (nullable id<ASLayoutable>)childForIndex:(NSUInteger)index;
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.h]: Use `childAtIndex:` and `setChild:atIndex:`

File: AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+  ASDisplayNodeAssert(self.isMutable, @"Cannot set properties when layout spec is not mutable");
+  
+  id<ASLayoutable> layoutable = child ? [self layoutableToAddFromLayoutable:child] : [ASNullLayoutSpec null];
+  
+  if (child) {
+    if (_childrenArray.count < index) {
+      // Fill up the array with null objects until the index
+      NSInteger i = _childrenArray.count;
+      while (i < index) {
+        _childrenArray[i] = [ASNullLayoutSpec null];
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm]: Have you looked at `NSPointerArray`? It's kind of kooky but it might be perfect for this.

Comment (maicki) [inline AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm]: Removed `NSNullLayoutSpec` and moved to `NSPointerArray`

File: AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.h (hunk 7 above)
Comment (maicki) [inline AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.h]: As it's only used in subclasses we should support for now until we know what the impact of the removal would have on already written code

File: AsyncDisplayKit/Layout/ASLayoutSpec.mm (hunk 5 above)
Comment (maicki) [inline AsyncDisplayKit/Layout/ASLayoutSpec.mm]: Cannot do anymore with `NSPointerArray`

File: AsyncDisplayKit/Layout/ASLayoutSpec.mm (hunk 6 above)
Comment (maicki) [inline AsyncDisplayKit/Layout/ASLayoutSpec.mm]: As it's only used in subclasses we should support for now until we know what the impact of the removal would have on already written code

File: AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.h (hunk 7 above)
Comment (maicki) [inline AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.h]: Changed

**Comment by maicki** [comment]
@Adlai-Holler ready for review

File: AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+      }
+    }
+  }
+  
+  // Replace object at the given index with the layoutable
+  _childrenArray[index] = layoutable;
+  
+  // TODO: Should we propagate up the layoutable at it could happen that multiple children will propagated up their
+  //       layout options and one child will overwrite values from another child
+  // [self propagateUpLayoutable:finalLayoutable];
```
Comment (appleguy) [inline AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm]: @maicki I guess this was commented out before?

File: AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+- (id<ASLayoutable>)childAtIndex:(NSUInteger)index
+{
+  id<ASLayoutable> layoutable = nil;
+  if (index < _childrenArray.count) {
+    layoutable = _childrenArray[index];
+  }
+  
+  // Null layoutable should not be accessed
+  ASDisplayNodeAssert(layoutable != [ASNullLayoutSpec null], @"Access child at index without set a child at that index");
```
Comment (appleguy) [inline AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm]: @maicki Does this check have to change for the pointer array or?

File: AsyncDisplayKit/Layout/ASLayoutSpec.mm [hunk 11]
```diff
@@ -162,67 +114,63 @@ - (void)setParent:(id<ASLayoutable>)parent
...
-- (NSArray *)children
+- (NSUInteger)countByEnumeratingWithState:(NSFastEnumerationState *)state objects:(id __unsafe_unretained _Nullable [_Nonnull])buffer count:(NSUInteger)len
 {
-  std::vector<ASLayout *> children;
-  for (ASChildMap::iterator it = _children.begin(); it != _children.end(); ++it ) {
-    children.push_back(it->second);
-  }
-  
-  return [NSArray arrayWithObjects:&children[0] count:children.size()];
+  return [_childrenArray countByEnumeratingWithState:state objects:buffer count:len];
```
Comment (appleguy) [inline AsyncDisplayKit/Layout/ASLayoutSpec.mm]: Is there anything that protects us against mutation while iterating?  @maicki

File: AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm (hunk 9 above)
Comment (maicki) [inline AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm]: Yes

File: AsyncDisplayKit/Layout/ASLayoutSpec.mm (hunk 11 above)
Comment (maicki) [inline AsyncDisplayKit/Layout/ASLayoutSpec.mm]: `NSMutableArray` internally should throw an assertion if something is mutating while iterating

File: AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm (hunk 10 above)
Comment (maicki) [inline AsyncDisplayKit/Layout/ASLayoutSpec+Subclasses.mm]: No as we don't use a pointer array

**Pull request merged**